index.ts

```diff
@@ -1,35 +1,65 @@
 export class OatyArray {
-  private _original: object[]
-  private _keys: string[]
-  private _transposed: object = {}
+  private keys: string[] | undefined
+  private original: object[] = []
+  private transposed: { [key: string]: any } = {}
+
+  constructor(options?: { data?: object[], keys?: string[] }) {
+    if (options !== undefined) {
+      if (options.data !== undefined) { this.push(...options.data) }
+      if (options.keys !== undefined) { this.keys = options.keys }
+    }
+  }
 
   get length(): number {
-    return this._original.length
+    return this.original.length
   }
 
-  constructor(input: object[], keys: string[]) {
-    this._original = input.map((x) => Object.assign({}, x))
-    this._keys = keys
+  public getKeys(): string[] | undefined {
+    return this.keys
+  }
 
-    this._transposed = transpose(this._keys, this._original)
+  public getOriginal(): object[] {
+    return this.original
   }
 
-  public get(keyName: string, keyValue: string): object[] {
-    return this._transposed[keyName][keyValue]
+  public getTransposed(): object[] {
+    return this.original
   }
 
-  public push(...items: object[]): number {
-    this._transposed = transpose(this._keys, items, this._transposed)
-    return this._original.push(...items)
+  public get(keyName: string, keyValue: string): object[] | undefined {
+    if (this.transposed === undefined
+     || this.transposed[keyName] === undefined) {
+      return undefined
+    }
+    return this.transposed[keyName][keyValue]
   }
-}
 
-const transpose = (keys: string[], items: object[], current?: object) => {
-  return keys.reduce((transposed, key) => {
-    transposed[key] = items.reduce((acc, item) => {
-      acc[item[key]] ? acc[item[key]].push(item) : acc[item[key]] = [item]
-      return acc
-    }, current || {})
-    return transposed
-  }, {})
+  public push(...data: object[]) {

```

Comment:
parameter name of `items` was more descriptive imo

Reply:
I wanted to standardise the name for it; you used `input` in the constructor.

Reply:
I felt like `input` was more descriptive for `constructor`, and `items` was more descriptive for `push`, but agree it might be a good idea to make them uniform. '`data`' seems generic though

Reply:
I've not tested it but my intention is to be able to push objects of different classes, so having a generic name is probably better.

Reply:
but in its current state it only accepts `object[]` - sort of makes more sense to rename when it needs a more accurate name, rather than in anticipation of a feature

Reply:
I couldn't think of anything to rename it to where it made sense in both contexts so I left it as `data`.